pkg/ntpdate/ntpdate.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ntpdate
+
+import (
+	"bufio"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/beevik/ntp"
+
+	"github.com/u-root/u-root/pkg/rtc"
+)
+
+type debugFn func(string, ...interface{})
+
+func parseServers(r *bufio.Reader, debug debugFn) []string {
+	var uri []string
+	var l string
+	var err error
+
+	debug("Reading config file")
+	for err == nil {
+		// This handles the case where the last line doesn't end in \n
+		l, err = r.ReadString('\n')
+		debug("%v", l)
+		if w := strings.Fields(l); len(w) > 1 && w[0] == "server" {
+			// We look only for the server lines, we ignore options like iburst
+			// TODO(ganshun): figure out what options we want to support.
+			uri = append(uri, w[1])
+		}
+	}
+
+	return uri
+}
+
+func getTime(servers []string, debug debugFn) (time.Time, string, error) {
+	for _, s := range servers {
+		debug("Getting time from %v", s)
+		t, err := ntp.Time(s)
+		if err == nil {
+			// Right now we return on the first valid time.
+			// We can implement better heuristics here.
+			debug("Got time %v", t)
+			return t, s, nil
+		}
+		debug("Error getting time from %s: %v", s, err)
+	}
+
+	return time.Time{}, "", fmt.Errorf("unable to get any time from servers %v", servers)
+}
+
+// SetTime sets system and optionally RTC time from NTP servers specified in sersers or the config file.
+func SetTime(servers []string, config string, fallback string, setRTC bool, verbose bool) error {
+	servers = servers[:]
+
+	debug := func(string, ...interface{}) {}
+	if verbose {
+		debug = log.Printf
+	}
+
+	if config != "" {
+		debug("Reading NTP servers from config file: %v", config)
+		f, err := os.Open(config)
+		if err == nil {
+			defer f.Close()
+			configServers := parseServers(bufio.NewReader(f), debug)
+			debug("Found %v servers", len(configServers))
+			servers = append(servers, configServers...)
+		} else {
+			debug("Unable to open config file: %v", err)
+		}
+	}
+
+	if len(servers) == 0 && fallback != "" {
+		debug("No servers provided, falling back to %v", fallback)
+		servers = append(servers, fallback)
+	}
+
+	if len(servers) == 0 {
+		return fmt.Errorf("no servers")
+	}
+
+	t, server, err := getTime(servers, debug)
+	if err != nil {
+		return fmt.Errorf("unable to get time: %w", err)
+	}
+
+	offset := time.Since(t).Seconds()
+	plus := ""
+	if offset > 0 {
+		plus = "+"
+	}
+	log.Printf("Setting time from %s (offset %s%f)...", server, plus, offset)
```

Comment:
This should be a call to [dD]ebug

Reply:
this is left verbose on purpose, ntpdate(8) outputs a similar message even in non-verbose mode:

```
$ sudo ntpdate time.google.com
10 Aug 19:59:39 ntpdate[518897]: adjust time server 216.239.35.0 offset -0.000991 sec
```

Reply:
isn't there some way to not have calls to log, however? packages that do random prints on their own are a frequent cause of bugs.